### PR TITLE
Resolve graphql queries through table functions

### PR DESCRIPTION
**Pull request by henneberger**
Moves the graphql query generation to more closely align with calcite. All graphql functions are resolved against calcite's table functions. Query generation was also revamped to remove some architectural drift. Parameter overloading can now happen. 

Notable Changes:
1. Table shadowing support has been removed except for distinct queries.
2. Parameter premutation can be disabled with an addArguments flag in the compiler config section

### CODE REVIEW

File: sqrl-planner/sqrl-planner-local/src/main/java/com/datasqrl/graphql/generate/ObjectTypeGenerator.java [hunk 1]
```diff
@@ -40,20 +43,33 @@ public Set<GraphQLType> visit(SqrlSchemaForInference schema, SchemaGeneratorCont
...
+        .fields(fields)
         .build();
   }
 
   @Override
   public GraphQLFieldDefinition visit(Column column, SchemaGeneratorContext context) {
     return GraphQLFieldDefinition.newFieldDefinition()
-        .name(column.getName().getDisplay())
-        .type(wrap(getOutputType(column.getType()), column.getType()))
+        .name(conformName(column.getName().getDisplay()))
```
Comment (henneberger) [inline sqrl-planner/sqrl-planner-local/src/main/java/com/datasqrl/graphql/generate/ObjectTypeGenerator.java:71]: This is a source of bugs. We'll need to either canonicalize the names or prevent them from being defined.

Comment (mbroecheler) [reply sqrl-planner/sqrl-planner-local/src/main/java/com/datasqrl/graphql/generate/ObjectTypeGenerator.java]: Aren't column names already canonicalized by the NameCanonicalizer?

File: sqrl-planner/sqrl-planner-local/src/main/java/com/datasqrl/graphql/generate/SchemaGenerator.java [hunk 2]
```diff
@@ -4,17 +4,108 @@
...
+    if (type instanceof GraphQLNonNull) {
+      return unbox(((GraphQLNonNull) type).getWrappedType());
+    } else if (type instanceof GraphQLList) {
+      return unbox(((GraphQLList) type).getWrappedType());
+    }
+    return type;
+  }
+
+  public static boolean isValidGraphQLName(String name) {
+    return Pattern.matches("[_A-Za-z][_0-9A-Za-z]*", name);
```
Comment (mbroecheler) [inline sqrl-planner/sqrl-planner-local/src/main/java/com/datasqrl/graphql/generate/SchemaGenerator.java:106]: Is there no such method in the GraphQL library?

File: sqrl-planner/sqrl-planner-local/src/main/java/com/datasqrl/graphql/inference/GraphQLSchemaGenerator.java [hunk 3]
```diff
@@ -0,0 +1,305 @@
+//package com.datasqrl.graphql.inference;
```
Comment (mbroecheler) [inline sqrl-planner/sqrl-planner-local/src/main/java/com/datasqrl/graphql/inference/GraphQLSchemaGenerator.java:1]: Does this need to be removed or is this added for later use?

File: sqrl-planner/sqrl-planner-local/src/main/java/com/datasqrl/graphql/inference/SchemaBuilder.java [hunk 4]
```diff
@@ -152,135 +164,202 @@ public Coords visitInterfaceField(InferredInterfaceField field, Object context)
...
+      });
+
+      if (argumentSet.getQuery() instanceof ApiQueryBase) {
+        int size = ((ApiQueryBase) argumentSet.getQuery())
+            .getParameters().size();
+
+        Preconditions.checkState(dynamicParams.size() == size,
+            "Args not matching");
+      }
+      System.out.println(argumentSet);
```
Comment (mbroecheler) [inline sqrl-planner/sqrl-planner-local/src/main/java/com/datasqrl/graphql/inference/SchemaBuilder.java]: remove println

File: sqrl-planner/sqrl-planner-local/src/main/java/com/datasqrl/plan/local/generate/AbstractTableNamespaceObject.java [hunk 5]
```diff
@@ -107,15 +107,16 @@ public void registerScriptTable(ScriptTableDefinition tblDef, SqrlFramework fram
...
         ScriptRelationalTable scriptParentTable = (ScriptRelationalTable) parentTable;
 
-        String fromSysTable = framework.getSchema().getPathToSysTableMap().get(parentPath);
-        String toSysTable = framework.getSchema().getPathToSysTableMap().get(path);
-
         Pair<List<FunctionParameter>, SqlNode> pkWrapper = createPkWrapper(scriptParentTable, table);
-        Relationship relationship = new Relationship(path.getLast(), path,
-            fromSysTable, toSysTable, JoinType.CHILD,
+        int version = framework.getUniqueMacroInt().incrementAndGet();
+        String internalName = String.join(".", path.toStringList()) + "$"
```
Comment (mbroecheler) [inline sqrl-planner/sqrl-planner-local/src/main/java/com/datasqrl/plan/local/generate/AbstractTableNamespaceObject.java:112]: see above

File: sqrl-planner/sqrl-planner-local/src/main/java/com/datasqrl/plan/local/generate/AbstractTableNamespaceObject.java [hunk 6]
```diff
@@ -128,14 +129,15 @@ public Relationship createParent(SqrlFramework framework, NamePath path, ScriptR
       ScriptRelationalTable childScriptTable) {
     Pair<List<FunctionParameter>, SqlNode> pkWrapper = createPkWrapper(childScriptTable,
         parentScriptTable);
-    String fromSysTable = framework.getSchema().getPathToSysTableMap().get(path);
-    String toSysTable = framework.getSchema().getPathToSysTableMap().get(path.popLast());
-
+    NamePath relPath = path.concat(ReservedName.PARENT);
+    int version = framework.getUniqueMacroInt().incrementAndGet();
+    String internalName = String.join(".", relPath.toStringList()) + "$"
```
Comment (mbroecheler) [inline sqrl-planner/sqrl-planner-local/src/main/java/com/datasqrl/plan/local/generate/AbstractTableNamespaceObject.java:134]: same

File: sqrl-planner/sqrl-planner-local/src/test/java/com/datasqrl/plan/local/analyze/ResolveTest.java [hunk 7]
```diff
@@ -685,10 +688,10 @@ private void validateAccessTableFunction(String tableName, String baseTableName,
 
   public static <T extends SqrlTableMacro> Optional<T> getLatestTableFunction(
       SqrlSchema sqrlSchema, String tableName) {
-    Optional<SqlUserDefinedTableFunction> tableFunction = sqrlSchema.getSqrlFramework().getQueryPlanner().getTableFunction(
-        List.of(tableName));
-    return tableFunction
-        .map(f->(T) f.getFunction());
+
+    List<Function> tableFunction =
+        new ArrayList<>(sqrlSchema.getSqrlFramework().getSchema().getFunctions(String.join(".", tableName), false));
```
Comment (mbroecheler) [inline sqrl-planner/sqrl-planner-local/src/test/java/com/datasqrl/plan/local/analyze/ResolveTest.java:693]: same

File: sqrl-testing/sqrl-integration-tests/system.discover.table5772692987871506298.json [hunk 8]
```diff
@@ -0,0 +1,11 @@
+{
```
Comment (mbroecheler) [inline sqrl-testing/sqrl-integration-tests/system.discover.table5772692987871506298.json:1]: should this be removed?

**Review by mbroecheler — CHANGES_REQUESTED** [review]
Thanks for reworking this part. Architecturally, this seems a bit intertwined. You mentioned that you wanted to refactor the schema generation from the mapping part - this might be what's tripping me up here.
My 2 cents: let's get this into good enough shape for the workshop and then refactor this code together afterward.
Left some comments to DRY the code.

File: sqrl-planner/sqrl-common/src/main/java/com/datasqrl/calcite/function/SqrlTableMacro.java [hunk 9]
```diff
@@ -19,5 +22,21 @@ default Type getElementType(List<Object> list) {
...
-  NamePath getPath();
+  NamePath getFullPath();
+  NamePath getAbsolutePath();
+
+  String getInternalName();
+  String getDisplayName();
+
+  @Value
+  class Column {
+    String name;
```
[the remark this answers is not in the document]
Comment (henneberger) [reply sqrl-planner/sqrl-common/src/main/java/com/datasqrl/calcite/function/SqrlTableMacro.java]: Dead code

File: sqrl-planner/sqrl-common/src/main/java/com/datasqrl/calcite/plan/ScriptPlanner.java [hunk 10]
```diff
@@ -126,15 +129,19 @@ public Void visit(SqrlAssignment assignment, Void context) {
...
-      String fromSysTable = planner.getSchema().getPathToSysTableMap().get(fromTable);
-      String toSysTable = planner.getSchema().getPathToSysTableMap().get(toTable);
+//      NamePath fromTable = path.popLast();
+      NamePath toTable = isASqrl.get(isASqrl.size()-1).getAbsolutePath();
+//      String fromSysTable = planner.getSchema().getPathToSysTableMap().get(fromTable);
+//      String toSysTable = planner.getSchema().getPathToSysTableMap().get(toTable);
       Supplier<RelNode> nodeSupplier = ()->expanded;
 
+      int version = framework.getUniqueMacroInt().incrementAndGet();
+      String internalName = String.join(".", path.toStringList()) + "$"
```
[the remark this answers is not in the document]
Comment (henneberger) [reply sqrl-planner/sqrl-common/src/main/java/com/datasqrl/calcite/plan/ScriptPlanner.java]: This was already refactored out. Removing.

File: sqrl-planner/sqrl-planner-local/src/main/java/com/datasqrl/graphql/generate/SchemaGenerator.java (hunk 2 above)
Comment (henneberger) [reply sqrl-planner/sqrl-planner-local/src/main/java/com/datasqrl/graphql/generate/SchemaGenerator.java:106]: No. It's a private variable. I copied it out of there.

**Review by mbroecheler — APPROVED** [review]
Looks good. Some minor comments we already discussed.

File: sqrl-planner/sqrl-common/src/main/java/com/datasqrl/function/SqrlFunctionParameter.java [hunk 11]
```diff
@@ -38,4 +40,8 @@ public RelDataType getType(RelDataTypeFactory relDataTypeFactory) {
   public boolean isOptional() {
     return defaultValue.isPresent();
   }
+
+  public String getVariableName() {
+    return name.charAt(0) == VARIABLE_PREFIX.getCanonical().charAt(0) ? name.substring(1) : name;
```
[the remark this answers is not in the document]
Comment (henneberger) [reply sqrl-planner/sqrl-common/src/main/java/com/datasqrl/function/SqrlFunctionParameter.java:45]: Ideally this should be inverted, the name stored is the one without the prefix, but it requires a parser change. I'd rather deal with that when we move to a full calcite parser. Also, this whole class should probably go in favor of standard function parameters and bookkeep the internal parameters in the table function.

**Pull request merged**